openssl/src/dh/mod.rs

```diff
@@ -0,0 +1,97 @@
+use ffi;
+use ssl::error::SslError;
+use bn::BigNum;
+use std::mem;
+use std::ptr;
+
+pub struct DH(*mut ffi::DH);
+
+impl DH {
+    pub fn from_params(p: BigNum, g: BigNum, q: BigNum) -> Result<DH, SslError> {
+        let dh = unsafe { ffi::DH_new_from_params(p.raw(), g.raw(), q.raw()) };
+        if dh == ptr::null_mut() {
+            return Err(SslError::get());
+        }
+        mem::forget(p);
+        mem::forget(g);
+        mem::forget(q);
+        Ok(DH(dh))
+    }
+
+    #[cfg(feature = "rfc5114")]
+    pub fn get_1024_160() -> Result<DH, SslError> {
+        let dh = unsafe { ffi::DH_get_1024_160() };
+        if dh == ptr::null_mut() {
```

Comment:
The  purpose of `from_params` is to allow reimplementing `get_2048_256()` on older OpenSSL versions. For custom parameters, reading the parameters from a file would indeed be a good addition. I'll submit a new PR for this.